### PR TITLE
forward client javascript component

### LINKED ISSUE #209: Create a didcomm forward client javascript client component.
https://github.com/ADORSYS-GIS/didcomm-messaging-clients-utilities

### PR DESCRIPTION
**Pull request by chendiblessing**

### CODE REVIEW

File: libs/did-resolver-lib/src/resolver.ts [hunk 1]
```diff
@@ -10,16 +10,35 @@ type Purpose =
   | 'Service';
 
 export default class PeerDIDResolver implements DIDResolver {
+  private diddocs: DIDDoc[];
+
+  constructor(diddocs: DIDDoc[] = []) {
+    this.diddocs = diddocs;
+  }
```
Comment (Hermann-Core) [inline libs/did-resolver-lib/src/resolver.ts]: why do you have an array of DID documents here?

Comment (chendiblessing) [reply libs/did-resolver-lib/src/resolver.ts]: i just merged the resolver in my branch to be able to use it i did not implement it

File: libs/message-handler/src/protocols/forward-client.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  ExampleDIDResolver,
+  ExampleSecretsResolver,
+} from '../../../did-resolver-lib/src/ExampleDIDResolver';
+import { FROM } from '../did_doc/client';
+import { ROUTING } from '../shared_data/message_types';
+import { CLIENT_SECRETS } from '../secrets/client';
+import { MEDIATOR_ENDPOINT } from '../shared_data/endpoints';
+import { CONTENT_TYPE } from '../shared_data/constant';
+
+export function buildMessage(to: string[], message: unknown): Message {
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts]: why do you use the unknown type here?

Comment (Christiantyemele) [reply libs/message-handler/src/protocols/forward-client.ts]: > why do you use the unknown type here?

Its because the message could be of any type, Array, string, etc but using any will make eslint complain

Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts]: > Its because the message could be of any type, Array, string, etc but using any will make eslint complain

You need a way to handle various input types within the function. If the function can accept any type of input, it should include logic to process these variations accordingly. It's not clear how your function addresses this.

**Review by Christiantyemele — CHANGES_REQUESTED** [review]
please undo all the changes in the did-resolver-lib

File: libs/message-handler/src/protocols/forward-client.ts (hunk 2 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts]: because the message can be of any type

Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts]: > because the message can be of any type

Read my previous comment again please.

**Review by Christiantyemele — APPROVED** [review]

**Review by Christiantyemele — CHANGES_REQUESTED** [review]
please go through the reviews and adjust them to in your issue they are repeating in all the clients

File: libs/message-handler/src/protocols/forward-client.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import {
+  DIDDoc,
+  DIDResolver,
+  IMessage,
+  Message,
+  SecretsResolver,
+} from 'didcomm';
+import { v4 as uuidv4 } from 'uuid';
+import axios from 'axios';
+import PeerDIDResolver from '../../../did-resolver-lib/src/resolver';
```
Comment (Christiantyemele) [inline libs/message-handler/src/protocols/forward-client.ts]: to import this as a package without specifying the part follow this steps
add this to did-resolver-libs/src/index.ts
```
export {
  ExampleDIDResolver,
  ExampleSecretsResolver,
} from './ExampleDIDResolver';
export { default as PeerDIDResolver } from './resolver';
```
then in the package.json of the did-resolver-lib add this
```
  "main": "src/index.ts" 
  ```
  then in the message-handler package.json add
  ```"did-resolver-lib": "*"```

File: libs/message-handler/src/app.tsx [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// src/main.tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+ReactDOM.render(
+  <React.StrictMode>
+    <h1>Message Handler</h1>
+  </React.StrictMode>,
+  document.getElementById('root'), // Ensure you have a corresponding HTML element with this ID
+);
```
Comment (Christiantyemele) [inline libs/message-handler/src/app.tsx]: what is this file for ? its not important remove it

File: libs/message-handler/tsconfig.app.json [hunk 5]
```diff
@@ -20,5 +21,8 @@
     "noUnusedParameters": true,
     "noFallthroughCasesInSwitch": true
   },
-  "include": ["src"]
+  "include": [
+    "src",
+    "../did-resolver-lib/src/forward.test.ts/src/forward.test.ts"
```
Comment (Christiantyemele) [inline libs/message-handler/tsconfig.app.json]:
```suggestion
    "src"
```

File: libs/message-handler/tsconfig.node.json [hunk 6]
```diff
@@ -18,5 +19,8 @@
     "noUnusedParameters": true,
     "noFallthroughCasesInSwitch": true
   },
-  "include": ["src"]
+  "include": [
+    "src",
+    "../did-resolver-lib/src/forward.test.ts/src/forward.test.ts"
+  ]
```
Comment (Christiantyemele) [inline libs/message-handler/tsconfig.node.json]:
```suggestion
    "src"
  ]
```

**Review by Christiantyemele — CHANGES_REQUESTED** [review]

File: libs/message-handler/src/protocols/forward-client.ts (hunk 3 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts]: it ia already done

File: libs/message-handler/src/app.tsx (hunk 4 above)
Comment (chendiblessing) [reply libs/message-handler/src/app.tsx]: done

**Review by Christiantyemele — APPROVED** [review]

File: libs/message-handler/src/protocols/forward-client.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    id: uuidv4(),
+    typ: 'application/didcomm-plain+json',
+    type: ROUTING,
+    body: {},
+    from: FROM,
+    to: to,
+    attachments: [
+      {
+        data: {
+          json: { data: message },
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts:32]: So the only valid attachment here is a string message, right?

File: libs/message-handler/src/protocols/forward-client.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        },
+        id: uuidv4(),
+        description: 'example',
+        media_type: 'application/didcomm-encrypted+json',
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts:36]: Are these fields necessary?

File: libs/message-handler/src/protocols/forward-client.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      to[0],
+      FROM,
+      null,
+      did_resolver,
+      secret_resolver,
+      {
+        forward: true,
+      },
+    );
+    return result[0];
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts:74]: Is `result` an array?

File: libs/message-handler/src/protocols/forward-client.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    mediator_didoc as DIDDoc,
+  ]);
+  const secret_resolver: SecretsResolver = new ExampleSecretsResolver(
+    CLIENT_SECRETS,
+  );
+
+  try {
+    const result = await msg.pack_encrypted(
+      to[0],
+      FROM,
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts:66]: I don't think the `FROM` header is not necessary for the forward message. From the spec:
> The only message in this protocol is the forward message. A simple and common version of a forward > > message might look like this:
>
> {
>    "type": "https://didcomm.org/routing/2.0/forward",
>    "id": "abc123xyz456",
>    "to": ["did:example:mediator"],
>    "expires_time": 1516385931,
>    "body":{
>        "next": "did:foo:1234abcd"
>    },
>    "attachments": [
>        // The payload(s) to be forwarded
>    ]
> }

File: libs/message-handler/src/protocols/forward-client.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    throw new Error(error as string);
+  }
+}
+export async function sendRequest(packed_msg: string) {
+  try {
+    await axios.post(MEDIATOR_ENDPOINT, packed_msg, {
+      headers: {
+        'Content-type': CONTENT_TYPE,
+      },
+    });
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts]: Please where do you handle the response that you get from the mediator?

File: libs/message-handler/src/tests/forward.test.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  test('should forward a message successfully', async () => {
+    const to = [
+      'did:peer:2.Vz6Mkf6r1uMJwoRAbzkuyj2RwPusdZhWSPeEknnTcKv2C2EN7.Ez6LSgbP4b3y8HVWG6C73WF2zLbzjDAPXjc33P2VfnVVHE347.SeyJpZCI6IiNkaWRjb21tIiwicyI6eyJhIjpbImRpZGNvbW0vdjIiXSwiciI6W10sInVyaSI6Imh0dHA6Ly9hbGljZS1tZWRpYXRvci5jb20ifSwidCI6ImRtIn0',
+    ];
+    const message = 'Hello is a test message to forward';
+    const msg = buildMessage(to, message);
+    const packmsg = pack_encrypt(msg, to);
+    expect(packmsg).not.toBeNull();
+  });
+});
```
Comment (Hermann-Core) [inline libs/message-handler/src/tests/forward.test.ts:24]: Could you include additional test scenarios to ensure that your code behaves as expected, even in non-ideal cases?

**Review by Hermann-Core — CHANGES_REQUESTED** [review]

File: libs/message-handler/src/protocols/forward-client.ts (hunk 7 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:32]: yes

File: libs/message-handler/src/protocols/forward-client.ts (hunk 8 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:36]: yes, The description and media_type fields are included to provide additional context about the attachment

File: libs/message-handler/src/protocols/forward-client.ts (hunk 9 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:74]: yes the result is an array

File: libs/message-handler/src/tests/forward.test.ts (hunk 12 above)
Comment (chendiblessing) [reply libs/message-handler/src/tests/forward.test.ts:24]: with the test i wrote if i should change the did to an invalid did the test will fail

File: libs/message-handler/src/protocols/forward-client.ts (hunk 8 above)
Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts:36]: I don't think they are important. But there will be no problem if we keep them here.

File: libs/message-handler/src/protocols/forward-client.ts (hunk 9 above)
Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts:74]: Given the signature of the called function, it returns a `string`, so I don't think it is correct. Can you show the signature of the function here?

**Edit**: Given the function's signature, it returns a tuple containing the message string and some metadata. Accessing result[0] will return the first element of the tuple, which is the message string. So it is ok. But result is not an array but a tuple.

File: libs/message-handler/src/tests/forward.test.ts (hunk 12 above)
Comment (Hermann-Core) [reply libs/message-handler/src/tests/forward.test.ts:24]: Then you assert the error.

File: libs/message-handler/src/protocols/forward-client.ts (hunk 10 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:66]: the from is necessary

Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts:66]: Could you link the part of the spec which stated that?

File: libs/message-handler/src/protocols/forward-client.ts (hunk 7 above)
Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts:32]: Please read the spec again. The attachments of the forward message is another jwe message for the recipient wrapped in a forward message envelope.

File: libs/message-handler/src/protocols/forward-client.ts (hunk 10 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:66]: image

File: libs/message-handler/src/protocols/forward-client.ts (hunk 7 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:32]: please can you rephrase the statement

File: libs/message-handler/src/protocols/forward-client.ts (hunk 8 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:36]: okay

File: libs/message-handler/src/tests/forward.test.ts (hunk 12 above)
Comment (chendiblessing) [reply libs/message-handler/src/tests/forward.test.ts:24]: i have added a test suite for that

File: libs/message-handler/src/protocols/forward-client.ts (hunk 11 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts]: in the forward protocol you don't get a response from the mediator

**Review by Tekum-Emmanuella — APPROVED** [review]

File: libs/message-handler/src/protocols/forward-client.ts (hunk 11 above)
Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts]: You always receive a response. How can you be sure that everything has proceeded correctly?

File: libs/message-handler/src/protocols/forward-client.ts (hunk 7 above)
Comment (Hermann-Core) [reply libs/message-handler/src/protocols/forward-client.ts:32]: Read this [message](https://identity.foundation/didcomm-messaging/spec/#messages) section from the spec.

File: libs/message-handler/src/protocols/forward-client.ts (hunk 11 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts]: okay i will add that

File: libs/message-handler/src/protocols/forward-client.ts (hunk 9 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:74]: okay thank you

Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts:74]: okay

File: libs/message-handler/src/protocols/forward-client.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    const response = await axios.post(MEDIATOR_ENDPOINT, packed_msg, {
+      headers: {
+        'Content-type': CONTENT_TYPE,
+      },
+    });
+
+    if (response.status >= 200 && response.status < 300) {
+      return response.data;
+    } else {
+      throw new Error(`Unexpected response status: ${response.status}`);
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts]: In case of error, you need to know what specific error message was sent by the mediator to get more details about the error rather than just the status error.

File: libs/message-handler/src/protocols/forward-client.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+export async function sendRequest(packed_msg: string) {
+  try {
+    const response = await axios.post(MEDIATOR_ENDPOINT, packed_msg, {
+      headers: {
+        'Content-type': CONTENT_TYPE,
+      },
+    });
+
+    if (response.status >= 200 && response.status < 300) {
+      return response.data;
```
Comment (Hermann-Core) [inline libs/message-handler/src/protocols/forward-client.ts]: In the event of success, do you know the specific response you receive from the mediator? How will you handle that response? I believe you will receive nothing more than a status code.

**Review by Hermann-Core — CHANGES_REQUESTED** [review]

File: libs/message-handler/src/protocols/forward-client.ts (hunk 13 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts]: done

File: libs/message-handler/src/protocols/forward-client.ts (hunk 14 above)
Comment (chendiblessing) [reply libs/message-handler/src/protocols/forward-client.ts]: done

File: libs/message-handler/src/placeholder.test.ts [hunk 15]
```diff
@@ -13,7 +13,6 @@ describe('didcomm', () => {
       expires_time: 1516385931,
       body: { messagespecificattribute: 'and its value' },
     });
-    const val = new Message(msg.as_value());
-    return val;
+    expect(msg.as_value()).toBeDefined();
```
Comment (Hermann-Core) [inline libs/message-handler/src/placeholder.test.ts:16]: what does this means?

**Review by Hermann-Core — APPROVED** [review]

File: libs/message-handler/src/placeholder.test.ts (hunk 15 above)
Comment (chendiblessing) [reply libs/message-handler/src/placeholder.test.ts:16]: i did not make changes on that branch i only merged the main branch
